test/bijectors/ordered.jl

```diff
@@ -17,23 +37,80 @@ using LinearAlgebra
+        # positive supports
+        product_distribution(fill(LogNormal(), 5)),
+        product_distribution(fill(InverseGamma(2, 3), 5)),
+        # negative supports
+        product_distribution(fill(-1 * InverseGamma(2, 3), 5)),
+        # bounded supports
+        product_distribution(fill(Uniform(1, 2), 5)),
+        product_distribution(fill(Beta(), 5)),
+        # different transformations
+        product_distribution(fill(transformed(InverseGamma(2, 3), Bijectors.Scale(3)), 5)),
+        product_distribution(fill(transformed(InverseGamma(2, 3), Bijectors.Scale(-3)), 5)),
+        product_distribution(fill(transformed(InverseGamma(2, 3), Bijectors.Shift(3)), 5)),
+        product_distribution(fill(transformed(InverseGamma(2, 3), Bijectors.Shift(-3)), 5)),
     ]
         d_ordered = ordered(d)
-        @test d_ordered isa Bijectors.TransformedDistribution
+        @test d_ordered isa Bijectors.OrderedDistribution
         @test d_ordered.dist === d
-        @test d_ordered.transform isa OrderedBijector
-        y = randn(5)
-        x = inverse(bijector(d_ordered))(y)
-        @test issorted(x)
+        num_tries = 100
+        for _ in 1:num_tries
+            y = rand(d)
+            x = inverse(bijector(d_ordered))(y)
+            @test issorted(x)
+            @test isfinite(logpdf(d_ordered, x))
+        end
+        # Check that `logpdf` correctly identifies out-of-support values.
+        @test !isfinite(logpdf(d_ordered, sort(rand(d); rev=true)))
     end
 
     @testset "non-identity bijector is not supported" begin
         d = Dirichlet(ones(5))
         @test_throws ArgumentError ordered(d)
     end
+
+    @testset "correctness" begin
+        num_samples = 10_000
+        num_adapts = 1_000
+        k = 2
+        @testset "$(typeof(dist))" for dist in [
+            MvNormal(1:k, Diagonal(1:k)),
+            product_distribution(fill(truncated(Normal(); lower=0), k)),
+        ]
+            prob = OrderedTestProblem(dist)
+            sampler = AdvancedHMC.NUTS(0.8)
+            initial_params = rand(ordered(dist))
+            transitions = sample(
+                prob,
+                sampler,
+                num_samples;
+                initial_params=initial_params,
+                discard_initial=num_adapts,
+                n_adapts=num_adapts,
+            )
+            xs = mapreduce(hcat, transitions) do t
+                to_constrained(prob, t.z.θ)
+            end
+            @test all(issorted, eachcol(xs))
+            # Should result in the same as rejection sampling.
+            dist_ordered = ordered(dist)
+            xs_true = rand(dist_ordered, num_samples)
+            @test all(issorted, eachcol(xs_true))
+
+            qs_true = mapslices(xs_true; dims=2) do xs
+                quantile(xs, [0.25, 0.5, 0.75])
+            end
+            qs = mapslices(xs; dims=2) do xs
+                quantile(xs, [0.25, 0.5, 0.75])
+            end
+
+            @info "quantiles" qs_true qs
```

Comment:
@sethaxen this is where I'm seeing a clear discrepancy still :confused:

Reply:
Ah, didn't see this. The issue looks to be that `rand!` is wrong for `OrderedDistribution`. Replacing it with the generic fallback defined below fixes the test.